### PR TITLE
config: simplify fetching widget configuration

### PR DESCRIPTION
**Pull request by ricardomv**
each widget has only his config object not the global object

we don't need to duplicate things in the config file if the user wants to remove a widget just remove the widget configuration from the file

### CODE REVIEW

**Comment by ricardomv** [comment]
i just noticed a problem in the last patch, waf has to rebuild the hole project every time you do `waf build`

**Comment by Lokaltog** [comment]
Does that happen for you with 77d14e92772c0a11d55da918695e9da15b551be1? I'm not able to reproduce that issue myself, here's my output when repeatedly doing waf build:

```
kim@regiontog ~/p/wkline> ./waf clean
'clean' finished successfully (0.012s)

kim@regiontog ~/p/wkline> ./waf configure build
Setting top to                           : /home/kim/projects/wkline 
Setting out to                           : /home/kim/projects/wkline/build 
Checking for 'gcc' (c compiler)          : /usr/bin/gcc 
Checking for program pkg-config          : /usr/bin/pkg-config 
Checking for pkg-config version >= '0.0.0' : yes 
Checking for 'gtk+-3.0'                    : yes 
Checking for 'glib-2.0 gmodule-2.0'        : yes 
Checking for 'webkitgtk-3.0'               : yes 
Checking for 'jansson'                     : yes 
Checking for 'alsa'                        : yes 
Checking for 'libcurl'                     : yes 
Checking for 'dbus-1 dbus-glib-1'          : yes 
Checking for 'libmpdclient'                : yes 
Checking for 'xcb-util xcb-ewmh xcb-icccm' : yes 
'configure' finished successfully (0.070s)
Waf: Entering directory `/home/kim/projects/wkline/build'
[ 1/24] c: src/util/config.c -> build/src/util/config.c.1.o
[ 2/24] c: src/util/copy_prop.c -> build/src/util/copy_prop.c.1.o
[ 3/24] c: src/util/log.c -> build/src/util/log.c.1.o
[ 4/24] c: src/widgets.c -> build/src/widgets.c.2.o
[ 5/24] c: src/widgets/volume.c -> build/src/widgets/volume.c.3.o
[ 6/24] c: src/util/curl.c -> build/src/util/curl.c.4.o
[ 7/24] c: src/widgets/external_ip.c -> build/src/widgets/external_ip.c.5.o
[ 8/24] c: src/widgets/weather.c -> build/src/widgets/weather.c.6.o
[ 9/24] c: src/util/dbus_helpers.c -> build/src/util/dbus_helpers.c.7.o
[10/24] c: src/widgets/battery.c -> build/src/widgets/battery.c.8.o
[11/24] c: src/widgets/notifications.c -> build/src/widgets/notifications.c.9.o
[12/24] c: src/widgets/now_playing_mpd.c -> build/src/widgets/now_playing_mpd.c.10.o
[13/24] c: src/widgets/desktops.c -> build/src/widgets/desktops.c.11.o
[14/24] c: src/widgets/window_title.c -> build/src/widgets/window_title.c.12.o
[15/24] c: src/wkline.c -> build/src/wkline.c.13.o
[16/24] cshlib: build/src/widgets/volume.c.3.o -> build/libwidget_volume.so
[17/24] cshlib: build/src/widgets/external_ip.c.5.o build/src/util/curl.c.4.o -> build/libwidget_external_ip.so
[18/24] cshlib: build/src/widgets/battery.c.8.o build/src/util/dbus_helpers.c.7.o -> build/libwidget_battery.so
[19/24] cshlib: build/src/widgets/weather.c.6.o build/src/util/curl.c.4.o -> build/libwidget_weather.so
[20/24] cshlib: build/src/widgets/notifications.c.9.o build/src/util/dbus_helpers.c.7.o -> build/libwidget_notifications.so
[21/24] cshlib: build/src/widgets/now_playing_mpd.c.10.o -> build/libwidget_now_playing_mpd.so
[22/24] cprogram: build/src/wkline.c.13.o build/src/util/config.c.1.o build/src/util/copy_prop.c.1.o build/src/util/log.c.1.o build/src/widgets.c.2.o -> build/wkline
[23/24] cshlib: build/src/widgets/window_title.c.12.o -> build/libwidget_window_title.so
[24/24] cshlib: build/src/widgets/desktops.c.11.o -> build/libwidget_desktops.so
Waf: Leaving directory `/home/kim/projects/wkline/build'
'build' finished successfully (1.043s)

kim@regiontog ~/p/wkline> ./waf build
Waf: Entering directory `/home/kim/projects/wkline/build'
Waf: Leaving directory `/home/kim/projects/wkline/build'
'build' finished successfully (0.044s)

kim@regiontog ~/p/wkline> ./waf build
Waf: Entering directory `/home/kim/projects/wkline/build'
Waf: Leaving directory `/home/kim/projects/wkline/build'
'build' finished successfully (0.042s)
```

**Comment by ricardomv** [comment]
no no my last patch in my `config` branch not `develop` i just fixed it patch comming

**Comment by ricardomv** [comment]
i had to change our config.c/h name again :S the header file was the same as config.h from waf defines

**Comment by Lokaltog** [comment]
Hm, that's annoying. Was that the reason for the unneeded rebuilds?

Also, I really don't want to get rid of the functionality provided by the enabled/ordered widget array. Do you know if jansson returns the JSON object in the same order it's stored in the config.json? If it does we could add an `"enabled": true/false` property to each widget and use that. I want to be able to easily disable a widget without having to delete the entire config object (e.g. for temporarily disabling a widget).

**Comment by ricardomv** [comment]
the rebuilds i think were because waf was passing the defines in the compile command and when they changed the all the files were compiled, now i write it to a file `config.h`, this is why i had to change the name of our `config.h`.
Isn't the order of widgets managed by the theme? and you can disable a widget by removing that "section" form this config file. to enable it again just copy the file from `/etc/xdg/wkline/config.json`

**Comment by Lokaltog** [comment]
I want the widget position to be user-configurable instead of locking it to the theme, which is why I want this to be a part of config.json.

I think having to remove the entire config block is a poor way of disabling a widget, what if you want to disable a widget with custom configuration, then you'll have to redo the configuration if you copy it from /etc/xdg/wkline/config.json. I have to insist on having an `enabled` option if we're going to only use the widgets config object, unless it's a lot of work to implement it.

File: wscript [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+	version = None
+	if os.path.exists(".git"):
+		version = "-git-"+os.popen("git rev-list --count HEAD").read().strip()
+		version += "."+os.popen("git rev-parse --short HEAD").read().strip()
+	return version
+
+def pre_build(ctx):
+	ctx.define('VERSION', VERSION+get_git_version())
+	ctx.define('BUILD_DATE', time.strftime("%c"))
+	ctx.write_config_header('src/config.h')
```
Comment (Lokaltog) [inline wscript:21]: Isn't it possible to pass these defines via command-line defines, like we do [here](https://github.com/ricardo-vieira/wkline/blob/config/wscript#L41)? Then we won't have to write this to a separate config file and we can keep the config.c/h name which I think is a bit better than wkconfig.c/h.

Comment (ricardomv) [inline wscript:21]: if you want them to change every time you build then no. if you want them to only change when you do `waf configure` then yes they can go with the other defines which i don't think is too bad

Comment (ricardomv) [inline wscript:21]: i just don't like the idea of passing defines through the compile command

Comment (Lokaltog) [inline wscript:21]: For our purposes I don't think it matters too much, I can't imagine many users _not_ using the waf build script, and I personally prefer system generated build-time defines to be passed via the command line as opposed to be cached in a file, as long as the defines aren't supposed to be changed by the user. If the user was supposed to be able to change these defines I'd agree that having the defines in a separate config.h file would make more sense.

Comment (ricardomv) [inline wscript:21]: can you fix it or you want me to change the last patch?

Comment (Lokaltog) [inline wscript:21]: I can fix it before I merge it, no problem.

**Comment by ricardomv** [comment]
this might be my last patch i'm starting to get busy with college stuff. the jscore stuff is a lot of work for just one person and i don't think it's worth the effort. i wish you luck with the project and i hope i helped more than i wasted your time.  farewell

**Comment by Lokaltog** [comment]
That's too bad. I've really appreciated your help, you've definitely helped bring the project to a much better state and your ideas and feedback have been very useful. I feel I've learned a lot of C from your contributions, so it's unfortunate if you're unable to continue working on the project. Anyways, you're welcome back anytime! Thanks a lot for all your help and good luck with college!

**Comment by Lokaltog** [comment]
I'll leave this open, I want to merge it but I'll implement an `enabled` option first.

**Comment by Lokaltog** [comment]
I remembered to set you as the author of the commit, I did some minor changes to implement the enabled flag, hope that's allright. There's some code repetition [here](https://github.com/Lokaltog/wkline/blob/1ce94d8e9b412b8acb18704acd60751348b6ae99/src/widgets.c#L101), do you have any ideas to how this can be improved? I thought about just allocating memory to the max amount of widgets but that could be wasteful if many widgets are disabled. Not sure what's best practice here, please let me know if you want me to change it.

**Comment by ricardomv** [comment]
yeah i would allocate the hole space. have you tested what happens if you don't have the configuration `enabled` in a widget?

**Comment by Lokaltog** [comment]
I did test it, if you don't have `enabled` in a widget it silently enables the widgets (enabled is the default behavior). It will only be disabled if you explicitly set `enabled: false`.

I'll rewrite it to allocate the whole space instead.

**Comment by ricardomv** [comment]
btw if you just removed the block you wouldn't have this problem. I know you said what if i have a different config, just make a backup of the file if you have non standard configurations. and in the future the user shouldn't even have to mess with the config file anyway. The `Arch Way™` makes more sense every day.

**Comment by Lokaltog** [comment]
Allright, I'll do it your way since it's simpler, and we can implement the extra property _if_ it's required in the future (it's not a big deal yet anyways).

**Comment by ricardomv** [comment]
:D i'm happy we agree this is the way to go

**Comment by Lokaltog** [comment]
Haha, I admit you were right, sorry that it took so long before I realized it :P

**Comment by ricardomv** [comment]
it doesn't matter if i am right or wrong. my job is easy i just limit to give ideas and defend  them and you need to choose witch ones are good or bad.

**Pull request closed without merging**